### ISSUE TITLE
How to authenticate kfp cli client for multi-user mode?

### ISSUE THREAD

**Issue by Tomcli** [kubeflow/pipelines#4377]
### What steps did you take:
When running Kubeflow pipelines on multi-user mode that has istio sidecars, the kfp CLI command no longer works:
```shell
tommyli$ kfp pipeline list
(403)
Reason: Forbidden
HTTP response headers: HTTPHeaderDict({'Cache-Control': 'no-cache, private', 'Content-Length': '19', 'Content-Type': 'text/plain', 'Date': 'Fri, 14 Aug 2020 23:35:56 GMT', 'Server': 'istio-envoy', 'X-Envoy-Decorator-Operation': 'ml-pipeline.kubeflow.svc.cluster.local:8888/*'})
HTTP response body: RBAC: access denied
```

### What happened:
We are not sure how to authenticate or add session token to the kfp CLI client.

### What did you expect to happen:
In single user mode, the above command should able to list the pipeline.
```shell
tommyli$ kfp pipeline list
# +--------------------------------------+-----------------------------------+---------------------------+
# | Pipeline ID                          | Name                              | Uploaded at               |
# +======================================+===================================+===========================+
# | 925415d5-18e9-4e08-b57f-3b06e3e54648 | echo_pipeline                     | 2020-07-02T17:55:30+00:00 |
# +--------------------------------------+-----------------------------------+---------------------------+
...
```

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?

Using the [Istio with Dex deployment for IBM Cloud](https://github.com/kubeflow/manifests/blob/master/kfdef/kfctl_ibm_dex_multi_user.yaml).

KFP version:  1.0

KFP SDK version:  1.0

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

**Comment by Bobgy** [comment]
How does the CLI get host? Looks like it's not using public host.

PRs welcomed on CLI, what do you think making some config file CLI can read from.

**Comment by yhwang** [comment]
I tried to investigated this issue these two days and here are my findings:
- kfp cli uses kubectl current-context to get the id-token, url and sends the request to query ml-pipeline API.
- since kfp cli uses kubectl context to call the API, it may be different from the kubeflow user that user want to act as.
  Currently, kfp uses some options to support google IAP. I haven't study those options yet. May need to make those options more generic and could be use for other kube cluster providers. Or add some new options.
- In the kubeflow side, it missed a proper `Policy` to tell istio how to authenticate the API request from kfp cli. A proper ServiceRoleBinding to grant the user to access the ml-pipeline-services is also needed.
 
I will start to work on the change that is needed in kfp cli and then fix the Policy and ServiceRoleBinding later.
Any suggestion?

**Comment by Tomcli** [comment]
Thanks @yhwang for the investigation. It sounds good to me and I know https://github.com/kubeflow/pipelines/pull/4386 is trying to add the namespace support for cli.

**Comment by yhwang** [comment]
@Tomcli thanks for the info. seems he may add userid header as well. But for this issue, we do need to add istio RBAC settings. Let me work on the kfp cli first.

**Comment by Bobgy** [comment]
Can I confirm, is it possible to let istio auto add the userid header? Because letting users specify it is a security concern, we cannot verify user identity. anyone can pretend to be any one else

**Comment by yhwang** [comment]
@Bobgy and @Tomcli do you know who is adequate to review my change?

**Comment by yhwang** [comment]
@Bobgy @Tomcli I guess with the change I made above, at least kfp cli can use `cookie` to simulate UX request and properly handle by ml-pipeline-ui and go the ml-pipeline.

However, I have another issue about the Python SDK user on the in-cluster jupyter notebook on kfp v1.1.0 . First, istio RBAC blocks the `kfp.Client()` calls from in-cluster jupyter notebook on **user's namespace**. I can manually fix that by adding servicerolebinding ( I think I can try to have a PR for this later, or if you know someone already works on this, please let me know). However, the API calls are still blocked by the ml-pipeline-api-server because of no `kubeflow-userid` header. Currently, seems the header is a must. should the logic be changed or somewhere the header could be injected?

**Comment by Tomcli** [comment]
@yhwang This is with the Python SDK where you use `kfp.Client()` right? We should open another issue to track it because this issue will be closed with your cli fixes.

**Comment by yhwang** [comment]
open issue #4440 for the Python SDK issue

**Comment by jonasdebeukelaer** [comment]
Hi, also running into this issue, thanks @yhwang for looking into a fix. 

Quick question though - how would one go about generating/finding the JWT to include in the header? This is quite unclear to me.

Also, is there any way to change from multi to single user mode in kubeflow? Currently unable to use the cli, and having to create python scripts that replicate it, using `kfp.Client` and specifying `client_id`/`client_secret` etc which isn't ideal.

**Comment by yhwang** [comment]
@jonasdebeukelaer 
> Quick question though - how would one go about generating/finding the JWT to include in the header? This is quite unclear to me.

If you are using OIDC for your authentication/authorization, those OIDC providers usually allow you to generate id-token through their token endpoint. Then istio can use the id-token to validate user identity (also through OIDC provider's token endpoint). I added istio authentication policy to perform JWT authentication for those requests coming from kfp cli. The key point is kfp cli needs to carry the id-token in the specific HTTP header. It works. However, the policy also apply to the requests coming from ml-pipeline-ui which **do not** have the JWT header and those requests are blocked. I am still trying to find a good solution for kfp cli use case.
